DependencyContextBuilder doesn't handle name conflicts between <Reference> and <PackageReference>

`dotnet build` the following project:

```XML
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>netstandard2.0</TargetFramework>
    <NETStandardImplicitPackageVersion>2.0.0-preview2-25331-01</NETStandardImplicitPackageVersion>
    <PreserveCompilationContext>true</PreserveCompilationContext>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="System.Threading.Thread" Version="4.3.0" />
  </ItemGroup>
</Project>
```

### Expected Results
Build succeeds

### Actual Results

```
C:\Program Files\dotnet\sdk\2.0.0-preview2-006327\Sdks\Microsoft.NET.Sdk\build\Microsoft.NET.Sdk.targets(102,5): error MSB4018: The "GenerateDepsFile" task failed unexpectedly.\r [F:\DotNetTest\CannotAd
dProperty\Repro\Repro.csproj]
C:\Program Files\dotnet\sdk\2.0.0-preview2-006327\Sdks\Microsoft.NET.Sdk\build\Microsoft.NET.Sdk.targets(102,5): error MSB4018: System.ArgumentException: Can not add property System.Threading.Thread to
Newtonsoft.Json.Linq.JObject. Property with the same name already exists on object.\r [F:\DotNetTest\CannotAddProperty\Repro\Repro.csproj]
C:\Program Files\dotnet\sdk\2.0.0-preview2-006327\Sdks\Microsoft.NET.Sdk\build\Microsoft.NET.Sdk.targets(102,5): error MSB4018:    at Newtonsoft.Json.Linq.JObject.ValidateToken(JToken o, JToken existing
)\r [F:\DotNetTest\CannotAddProperty\Repro\Repro.csproj]
C:\Program Files\dotnet\sdk\2.0.0-preview2-006327\Sdks\Microsoft.NET.Sdk\build\Microsoft.NET.Sdk.targets(102,5): error MSB4018:    at Newtonsoft.Json.Linq.JContainer.InsertItem(Int32 index, JToken item,
 Boolean skipParentCheck)\r [F:\DotNetTest\CannotAddProperty\Repro\Repro.csproj]
C:\Program Files\dotnet\sdk\2.0.0-preview2-006327\Sdks\Microsoft.NET.Sdk\build\Microsoft.NET.Sdk.targets(102,5): error MSB4018:    at Newtonsoft.Json.Linq.JObject.InsertItem(Int32 index, JToken item, Bo
olean skipParentCheck)\r [F:\DotNetTest\CannotAddProperty\Repro\Repro.csproj]
C:\Program Files\dotnet\sdk\2.0.0-preview2-006327\Sdks\Microsoft.NET.Sdk\build\Microsoft.NET.Sdk.targets(102,5): error MSB4018:    at Newtonsoft.Json.Linq.JContainer.AddInternal(Int32 index, Object cont
ent, Boolean skipParentCheck)\r [F:\DotNetTest\CannotAddProperty\Repro\Repro.csproj]
C:\Program Files\dotnet\sdk\2.0.0-preview2-006327\Sdks\Microsoft.NET.Sdk\build\Microsoft.NET.Sdk.targets(102,5): error MSB4018:    at Newtonsoft.Json.Linq.JContainer.AddInternal(Int32 index, Object cont
ent, Boolean skipParentCheck)\r [F:\DotNetTest\CannotAddProperty\Repro\Repro.csproj]
C:\Program Files\dotnet\sdk\2.0.0-preview2-006327\Sdks\Microsoft.NET.Sdk\build\Microsoft.NET.Sdk.targets(102,5): error MSB4018:    at Newtonsoft.Json.Linq.JContainer.Add(Object content)\r [F:\DotNetTest
\CannotAddProperty\Repro\Repro.csproj]
C:\Program Files\dotnet\sdk\2.0.0-preview2-006327\Sdks\Microsoft.NET.Sdk\build\Microsoft.NET.Sdk.targets(102,5): error MSB4018:    at Newtonsoft.Json.Linq.JObject..ctor(Object content)\r [F:\DotNetTest\
CannotAddProperty\Repro\Repro.csproj]
C:\Program Files\dotnet\sdk\2.0.0-preview2-006327\Sdks\Microsoft.NET.Sdk\build\Microsoft.NET.Sdk.targets(102,5): error MSB4018:    at Microsoft.Extensions.DependencyModel.DependencyContextWriter.AddDepe
ndencies(JObject libraryObject, IEnumerable`1 dependencies)\r [F:\DotNetTest\CannotAddProperty\Repro\Repro.csproj]
C:\Program Files\dotnet\sdk\2.0.0-preview2-006327\Sdks\Microsoft.NET.Sdk\build\Microsoft.NET.Sdk.targets(102,5): error MSB4018:    at Microsoft.Extensions.DependencyModel.DependencyContextWriter.WriteTa
rgetLibrary(Library library)\r [F:\DotNetTest\CannotAddProperty\Repro\Repro.csproj]
C:\Program Files\dotnet\sdk\2.0.0-preview2-006327\Sdks\Microsoft.NET.Sdk\build\Microsoft.NET.Sdk.targets(102,5): error MSB4018:    at Microsoft.Extensions.DependencyModel.DependencyContextWriter.<WriteT
arget>b__7_0(Library library)\r [F:\DotNetTest\CannotAddProperty\Repro\Repro.csproj]
C:\Program Files\dotnet\sdk\2.0.0-preview2-006327\Sdks\Microsoft.NET.Sdk\build\Microsoft.NET.Sdk.targets(102,5): error MSB4018:    at System.Linq.Enumerable.SelectArrayIterator`2.MoveNext()\r [F:\DotNet
Test\CannotAddProperty\Repro\Repro.csproj]
C:\Program Files\dotnet\sdk\2.0.0-preview2-006327\Sdks\Microsoft.NET.Sdk\build\Microsoft.NET.Sdk.targets(102,5): error MSB4018:    at Newtonsoft.Json.Linq.JContainer.AddInternal(Int32 index, Object cont
ent, Boolean skipParentCheck)\r [F:\DotNetTest\CannotAddProperty\Repro\Repro.csproj]
C:\Program Files\dotnet\sdk\2.0.0-preview2-006327\Sdks\Microsoft.NET.Sdk\build\Microsoft.NET.Sdk.targets(102,5): error MSB4018:    at Newtonsoft.Json.Linq.JContainer.Add(Object content)\r [F:\DotNetTest
\CannotAddProperty\Repro\Repro.csproj]
C:\Program Files\dotnet\sdk\2.0.0-preview2-006327\Sdks\Microsoft.NET.Sdk\build\Microsoft.NET.Sdk.targets(102,5): error MSB4018:    at Newtonsoft.Json.Linq.JObject..ctor(Object content)\r [F:\DotNetTest\
CannotAddProperty\Repro\Repro.csproj]
C:\Program Files\dotnet\sdk\2.0.0-preview2-006327\Sdks\Microsoft.NET.Sdk\build\Microsoft.NET.Sdk.targets(102,5): error MSB4018:    at Microsoft.Extensions.DependencyModel.DependencyContextWriter.WriteTa
rget(IReadOnlyList`1 libraries)\r [F:\DotNetTest\CannotAddProperty\Repro\Repro.csproj]
C:\Program Files\dotnet\sdk\2.0.0-preview2-006327\Sdks\Microsoft.NET.Sdk\build\Microsoft.NET.Sdk.targets(102,5): error MSB4018:    at Microsoft.Extensions.DependencyModel.DependencyContextWriter.WriteTa
rgets(DependencyContext context)\r [F:\DotNetTest\CannotAddProperty\Repro\Repro.csproj]
C:\Program Files\dotnet\sdk\2.0.0-preview2-006327\Sdks\Microsoft.NET.Sdk\build\Microsoft.NET.Sdk.targets(102,5): error MSB4018:    at Microsoft.Extensions.DependencyModel.DependencyContextWriter.Write(D
ependencyContext context)\r [F:\DotNetTest\CannotAddProperty\Repro\Repro.csproj]
C:\Program Files\dotnet\sdk\2.0.0-preview2-006327\Sdks\Microsoft.NET.Sdk\build\Microsoft.NET.Sdk.targets(102,5): error MSB4018:    at Microsoft.Extensions.DependencyModel.DependencyContextWriter.Write(D
ependencyContext context, Stream stream)\r [F:\DotNetTest\CannotAddProperty\Repro\Repro.csproj]
C:\Program Files\dotnet\sdk\2.0.0-preview2-006327\Sdks\Microsoft.NET.Sdk\build\Microsoft.NET.Sdk.targets(102,5): error MSB4018:    at Microsoft.NET.Build.Tasks.GenerateDepsFile.ExecuteCore()\r [F:\DotNe
tTest\CannotAddProperty\Repro\Repro.csproj]
C:\Program Files\dotnet\sdk\2.0.0-preview2-006327\Sdks\Microsoft.NET.Sdk\build\Microsoft.NET.Sdk.targets(102,5): error MSB4018:    at Microsoft.NET.Build.Tasks.TaskBase.Execute()\r [F:\DotNetTest\Cannot
AddProperty\Repro\Repro.csproj]
C:\Program Files\dotnet\sdk\2.0.0-preview2-006327\Sdks\Microsoft.NET.Sdk\build\Microsoft.NET.Sdk.targets(102,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.IT
askExecutionHost.Execute()\r [F:\DotNetTest\CannotAddProperty\Repro\Repro.csproj]
C:\Program Files\dotnet\sdk\2.0.0-preview2-006327\Sdks\Microsoft.NET.Sdk\build\Microsoft.NET.Sdk.targets(102,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__25.M
oveNext() [F:\DotNetTest\CannotAddProperty\Repro\Repro.csproj]
```

--- a/src/Tasks/Microsoft.NET.Build.Tasks.UnitTests/GivenADependencyContextBuilder.cs
+++ b/src/Tasks/Microsoft.NET.Build.Tasks.UnitTests/GivenADependencyContextBuilder.cs
@@ -1,8 +1,6 @@
 ﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
-using System.Collections.Generic;
-using System.IO;
 using FluentAssertions;
 using FluentAssertions.Json;
 using Microsoft.Build.Framework;
@@ -11,6 +9,9 @@ using Newtonsoft.Json;
 using Newtonsoft.Json.Linq;
 using NuGet.Frameworks;
 using NuGet.ProjectModel;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
 using Xunit;
 
 namespace Microsoft.NET.Build.Tasks.UnitTests
@@ -43,7 +44,7 @@ namespace Microsoft.NET.Build.Tasks.UnitTests
 
             IEnumerable<ReferenceInfo> directReferences =
                 ReferenceInfo.CreateDirectReferenceInfos(
-                    referencePaths ?? new ITaskItem[] { }, 
+                    referencePaths ?? new ITaskItem[] { },
                     referenceSatellitePaths ?? new ITaskItem[] { });
 
             ProjectContext projectContext = lockFile.CreateProjectContext(
@@ -85,19 +86,7 @@ namespace Microsoft.NET.Build.Tasks.UnitTests
         {
             get
             {
-                CompilationOptions compilationOptions = new CompilationOptions(
-                    defines: new[] { "DEBUG", "TRACE" },
-                    languageVersion: "6",
-                    platform: "x64",
-                    allowUnsafe: true,
-                    warningsAsErrors: false,
-                    optimize: null,
-                    keyFile: "../keyfile.snk",
-                    delaySign: null,
-                    publicSign: null,
-                    debugType: "portable",
-                    emitEntryPoint: true,
-                    generateXmlDocumentation: true);
+                CompilationOptions compilationOptions = CreateCompilationOptions();
 
                 ITaskItem[] dotnetNewSatelliteAssemblies = new ITaskItem[]
                 {
@@ -182,6 +171,121 @@ namespace Microsoft.NET.Build.Tasks.UnitTests
                     }
                 }
             }
+        }
+
+        [Fact]
+        public void ItDoesntCreateReferenceAssembliesWhenNoCompilationOptions()
+        {
+            DependencyContext dependencyContext = BuildDependencyContextWithReferenceAssemblies(useCompilationOptions: false);
+
+            dependencyContext.CompileLibraries.Should().BeEmpty();
+            dependencyContext
+                .RuntimeLibraries
+                .Should()
+                .NotContain(l => l.Type == "referenceassembly");
+            dependencyContext
+                .RuntimeLibraries
+                .SelectMany(l => l.Dependencies)
+                .Should()
+                .NotBeEmpty()
+                .And
+                .NotContain(d => d.Name == "System.NotConflicting")
+                .And
+                .NotContain(d => d.Name == "System.Collections.NonGeneric.Reference");
+        }
+
+        [Fact]
+        public void ItHandlesReferenceAndPackageReferenceNameCollisions()
+        {
+            DependencyContext dependencyContext = BuildDependencyContextWithReferenceAssemblies(useCompilationOptions: true);
+
+            dependencyContext.CompileLibraries.Should()
+                .Contain(c => c.Name == "System.NotConflicting" && c.Type == "referenceassembly");
+
+            // Note: System.Collections.NonGeneric is referenced in the lockfile, so DependencyContextBuilder
+            // appends ".Reference" to make it unique
+            dependencyContext.CompileLibraries.Should()
+                .Contain(c => c.Name == "System.Collections.NonGeneric.Reference" && c.Type == "referenceassembly");
+            dependencyContext.CompileLibraries.Should()
+                .Contain(c => c.Name == "System.Collections.NonGeneric.Reference.Reference" && c.Type == "referenceassembly");
+        }
+
+        private DependencyContext BuildDependencyContextWithReferenceAssemblies(bool useCompilationOptions)
+        {
+            string mainProjectName = "simple.dependencies";
+            LockFile lockFile = TestLockFiles.GetLockFile(mainProjectName);
+
+            SingleProjectInfo mainProject = SingleProjectInfo.Create(
+                "/usr/Path",
+                mainProjectName,
+                ".dll",
+                "1.0.0",
+                new ITaskItem[] { });
+
+            ITaskItem[] referencePaths = new ITaskItem[]
+            {
+                new MockTaskItem(
+                    "/usr/Path/System.NotConflicting.dll",
+                    new Dictionary<string, string>
+                    {
+                        { "CopyLocal", "false" },
+                        { "FusionName", "System.NotConflicting, Version=4.0.0.0, Culture=neutral, PublicKeyToken=null" },
+                        { "Version", "" },
+                    }),
+                new MockTaskItem(
+                    "/usr/Path/System.Collections.NonGeneric.dll",
+                    new Dictionary<string, string>
+                    {
+                        { "CopyLocal", "false" },
+                        { "FusionName", "System.Collections.NonGeneric, Version=4.0.0.0, Culture=neutral, PublicKeyToken=null" },
+                        { "Version", "" },
+                    }),
+                new MockTaskItem(
+                    "/usr/Path/System.Collections.NonGeneric.Reference.dll",
+                    new Dictionary<string, string>
+                    {
+                        { "CopyLocal", "false" },
+                        { "FusionName", "System.Collections.NonGeneric.Reference, Version=4.0.0.0, Culture=neutral, PublicKeyToken=null" },
+                        { "Version", "" },
+                    }),
+            };
+
+            ProjectContext projectContext = lockFile.CreateProjectContext(
+                FrameworkConstants.CommonFrameworks.NetCoreApp10,
+                runtime: null,
+                platformLibraryName: Constants.DefaultPlatformLibrary,
+                isSelfContained: false);
+
+            CompilationOptions compilationOptions = 
+                useCompilationOptions ? CreateCompilationOptions() :
+                null;
+
+            DependencyContext dependencyContext = new DependencyContextBuilder(mainProject, projectContext)
+                .WithReferenceAssemblies(ReferenceInfo.CreateReferenceInfos(referencePaths))
+                .WithCompilationOptions(compilationOptions)
+                .Build();
+
+            // ensure the DependencyContext can be written out successfully - it has no duplicate dependency names
+            Save(dependencyContext);
+
+            return dependencyContext;
+        }
+
+        private static CompilationOptions CreateCompilationOptions()
+        {
+            return new CompilationOptions(
+                    defines: new[] { "DEBUG", "TRACE" },
+                    languageVersion: "6",
+                    platform: "x64",
+                    allowUnsafe: true,
+                    warningsAsErrors: false,
+                    optimize: null,
+                    keyFile: "../keyfile.snk",
+                    delaySign: null,
+                    publicSign: null,
+                    debugType: "portable",
+                    emitEntryPoint: true,
+                    generateXmlDocumentation: true);
         }
     }
 }

--- a/src/Tasks/Microsoft.NET.Build.Tasks/DependencyContextBuilder.cs
+++ b/src/Tasks/Microsoft.NET.Build.Tasks/DependencyContextBuilder.cs
@@ -28,6 +28,8 @@ namespace Microsoft.NET.Build.Tasks
         private string _referenceAssembliesPath;
         private Dictionary<PackageIdentity, string> _filteredPackages;
         private bool _includeMainProjectInDepsFile = true;
+        private HashSet<string> _usedLibraryNames;
+        private Dictionary<ReferenceInfo, string> _referenceLibraryNames;
 
         public DependencyContextBuilder(SingleProjectInfo mainProjectInfo, ProjectContext projectContext)
         {
@@ -36,6 +38,42 @@ namespace Microsoft.NET.Build.Tasks
 
             // This resolver is only used for building file names, so that base path is not required.
             _versionFolderPathResolver = new VersionFolderPathResolver(rootPath: null);
+        }
+
+        /// <summary>
+        /// Keeps track of the Library names being used in the DependencyContext.
+        /// </summary>
+        /// <remarks>
+        /// Since `Reference` and `PackageReference` names can conflict, we need to ensure
+        /// each separate Library has a unique name. Since PackageReference names are guaranteed
+        /// to be unique amongst other PackageReferences, start with that set, and ensure
+        /// Reference names are unique amongst all.
+        /// </remarks>
+        private HashSet<string> UsedLibraryNames
+        {
+            get
+            {
+                if (_usedLibraryNames == null)
+                {
+                    _usedLibraryNames = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+                    _usedLibraryNames.AddRange(_projectContext.LockFile.Libraries.Select(l => l.Name));
+                }
+
+                return _usedLibraryNames;
+            }
+        }
+
+        private Dictionary<ReferenceInfo, string> ReferenceLibraryNames
+        {
+            get
+            {
+                if (_referenceLibraryNames == null)
+                {
+                    _referenceLibraryNames = new Dictionary<ReferenceInfo, string>();
+                }
+
+                return _referenceLibraryNames;
+            }
         }
 
         public DependencyContextBuilder WithMainProjectInDepsFile(bool includeMainProjectInDepsFile)
@@ -116,7 +154,8 @@ namespace Microsoft.NET.Build.Tasks
                     GetProjectRuntimeLibrary(
                         _mainProjectInfo,
                         _projectContext,
-                        dependencyLookup)
+                        dependencyLookup,
+                        includeCompilationLibraries)
                 });
             }
             runtimeLibraries = runtimeLibraries
@@ -133,7 +172,8 @@ namespace Microsoft.NET.Build.Tasks
                         GetProjectCompilationLibrary(
                             _mainProjectInfo,
                             _projectContext,
-                            dependencyLookup)
+                            dependencyLookup,
+                            includeCompilationLibraries)
                     });
                 }
 
@@ -183,7 +223,8 @@ namespace Microsoft.NET.Build.Tasks
 
         private List<Dependency> GetProjectDependencies(
             ProjectContext projectContext,
-            Dictionary<string, Dependency> dependencyLookup)
+            Dictionary<string, Dependency> dependencyLookup,
+            bool includeCompilationLibraries)
         {
             List<Dependency> dependencies = new List<Dependency>();
 
@@ -197,12 +238,17 @@ namespace Microsoft.NET.Build.Tasks
             }
 
             var referenceInfos = Enumerable.Concat(
-                _referenceAssemblies ?? Enumerable.Empty<ReferenceInfo>(),
+                includeCompilationLibraries && _referenceAssemblies != null ? 
+                    _referenceAssemblies : 
+                    Enumerable.Empty<ReferenceInfo>(),
                 _directReferences ?? Enumerable.Empty<ReferenceInfo>());
 
             foreach (ReferenceInfo referenceInfo in referenceInfos)
             {
-                dependencies.Add(new Dependency(referenceInfo.Name, referenceInfo.Version));
+                dependencies.Add(
+                    new Dependency(
+                        GetReferenceLibraryName(referenceInfo), 
+                        referenceInfo.Version));
             }
 
             return dependencies;
@@ -239,14 +285,16 @@ namespace Microsoft.NET.Build.Tasks
                 runtimeStoreManifestName: runtimeStoreManifestName,
                 serviceable: serviceable);
         }
+
         private RuntimeLibrary GetProjectRuntimeLibrary(
             SingleProjectInfo projectInfo,
             ProjectContext projectContext,
-            Dictionary<string, Dependency> dependencyLookup)
+            Dictionary<string, Dependency> dependencyLookup,
+            bool includeCompilationLibraries)
         {
             RuntimeAssetGroup[] runtimeAssemblyGroups = new[] { new RuntimeAssetGroup(string.Empty, projectInfo.OutputName) };
 
-            List<Dependency> dependencies = GetProjectDependencies(projectContext, dependencyLookup);
+            List<Dependency> dependencies = GetProjectDependencies(projectContext, dependencyLookup, includeCompilationLibraries);
 
             return CreateRuntimeLibrary(
                 type: "project",
@@ -263,9 +311,10 @@ namespace Microsoft.NET.Build.Tasks
         private CompilationLibrary GetProjectCompilationLibrary(
             SingleProjectInfo projectInfo,
             ProjectContext projectContext,
-            Dictionary<string, Dependency> dependencyLookup)
+            Dictionary<string, Dependency> dependencyLookup,
+            bool includeCompilationLibraries)
         {
-            List<Dependency> dependencies = GetProjectDependencies(projectContext, dependencyLookup);
+            List<Dependency> dependencies = GetProjectDependencies(projectContext, dependencyLookup, includeCompilationLibraries);
 
             return new CompilationLibrary(
                 type: "project",
@@ -457,7 +506,7 @@ namespace Microsoft.NET.Build.Tasks
             return _referenceAssemblies
                 ?.Select(r => new CompilationLibrary(
                     type: "referenceassembly",
-                    name: r.Name,
+                    name: GetReferenceLibraryName(r),
                     version: r.Version,
                     hash: string.Empty,
                     assemblies: new[] { ResolveFrameworkReferencePath(r.FullPath) },
@@ -485,7 +534,7 @@ namespace Microsoft.NET.Build.Tasks
             return _directReferences
                 ?.Select(r => CreateRuntimeLibrary(
                     type: "reference",
-                    name: r.Name,
+                    name: GetReferenceLibraryName(r),
                     version: r.Version,
                     hash: string.Empty,
                     runtimeAssemblyGroups: new[] { new RuntimeAssetGroup(string.Empty, r.FileName) },
@@ -502,7 +551,7 @@ namespace Microsoft.NET.Build.Tasks
             return _directReferences
                 ?.Select(r => new CompilationLibrary(
                     type: "reference",
-                    name: r.Name,
+                    name: GetReferenceLibraryName(r),
                     version: r.Version,
                     hash: string.Empty,
                     assemblies: new[] { r.FileName },
@@ -510,6 +559,38 @@ namespace Microsoft.NET.Build.Tasks
                     serviceable: false))
                 ??
                 Enumerable.Empty<CompilationLibrary>();
+        }
+
+        private string GetReferenceLibraryName(ReferenceInfo reference)
+        {
+            if (!ReferenceLibraryNames.TryGetValue(reference, out string name))
+            {
+                // Reference names can conflict with PackageReference names, so
+                // ensure that the Reference names are unique when creating libraries
+                name = GetUniqueReferenceName(reference.Name);
+
+                ReferenceLibraryNames.Add(reference, name);
+                UsedLibraryNames.Add(name);
+            }
+
+            return name;
+        }
+
+        private string GetUniqueReferenceName(string name)
+        {
+            if (UsedLibraryNames.Contains(name))
+            {
+                string startingName = $"{name}.Reference";
+                name = startingName;
+
+                int suffix = 1;
+                while (UsedLibraryNames.Contains(name))
+                {
+                    name = $"{startingName}{suffix++}";
+                }
+            }
+
+            return name;
         }
 
         private static IEnumerable<ResourceAssembly> CreateResourceAssemblies(IEnumerable<ResourceAssemblyInfo> resourceAssemblyInfos)

--- a/src/Tasks/Microsoft.NET.Build.Tasks/DependencyContextBuilder.cs
+++ b/src/Tasks/Microsoft.NET.Build.Tasks/DependencyContextBuilder.cs
@@ -55,8 +55,9 @@ namespace Microsoft.NET.Build.Tasks
             {
                 if (_usedLibraryNames == null)
                 {
-                    _usedLibraryNames = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
-                    _usedLibraryNames.AddRange(_projectContext.LockFile.Libraries.Select(l => l.Name));
+                    _usedLibraryNames = new HashSet<string>(
+                        _projectContext.LockFile.Libraries.Select(l => l.Name),
+                        StringComparer.OrdinalIgnoreCase);
                 }
 
                 return _usedLibraryNames;

--- a/test/Microsoft.NET.Publish.Tests/GivenThatWeWantToPreserveCompilationContext.cs
+++ b/test/Microsoft.NET.Publish.Tests/GivenThatWeWantToPreserveCompilationContext.cs
@@ -119,7 +119,7 @@ namespace Microsoft.NET.Publish.Tests
 
                         var systemCollectionsLibrary = dependencyContext
                             .CompileLibraries
-                            .FirstOrDefault(l => string.Equals(l.Name, "system.collections", StringComparison.OrdinalIgnoreCase));
+                            .FirstOrDefault(l => string.Equals(l.Name, "system.collections.reference", StringComparison.OrdinalIgnoreCase));
                         systemCollectionsLibrary.Assemblies.Count.Should().Be(1);
                         systemCollectionsLibrary.Assemblies[0].Should().Be(".NETFramework/v4.6/Facades/System.Collections.dll");
                     }

